### PR TITLE
RPMB emulation HMAC support

### PR DESCRIPTION
**Pull request by jforissier**
~~This PR is on top of https://github.com/OP-TEE/optee_client/pull/43, so please consider only the last commit, c74fbb2 ("RPMB: add HMAC to emulation mode").~~

### CODE REVIEW

**Comment by jforissier** [comment]
Copyright notice updated.

**Comment by jforissier** [comment]
Rebased since https://github.com/OP-TEE/optee_client/pull/43 has been merged. I also fixed the debug dump function.

File: tee-supplicant/src/rpmb.c [hunk 1]
```diff
@@ -356,35 +459,24 @@ static int ioctl_emu(int fd, unsigned long request, ...)
...
-			frm->write_counter = htonl(mem->write_counter);
-			memcpy(frm->nonce, mem->nonce, 16);
+			ioctl_emu_read_ctr(mem, frm);
 			break;
 
 		case RPMB_MSG_TYPE_REQ_AUTH_DATA_READ:
-			/* Read data */
-			DMSG("cmd->blocks = %d", cmd->blocks);
 			ioctl_emu_mem_transfer(mem, frm, cmd->blocks, 0);
-			memcpy(frm[cmd->blocks-1].nonce, mem->nonce, 16);
```
Comment (pascal-brand-st-dev) [inline tee-supplicant/src/rpmb.c]: we now copy the non on all blocks, not only the last one. I think it is ok, but I would like a confirmation.

File: tee-supplicant/src/sha2.c [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+/*
+ * Copyright (c) 2016, Linaro Limited
```
Comment (pascal-brand-st-dev) [inline tee-supplicant/src/sha2.c]: Linaro copyright to be included in the original one, without the whole license terms

File: tee-supplicant/src/sha2.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+/*
+ * Copyright (c) 2016, Linaro Limited
```
Comment (pascal-brand-st-dev) [inline tee-supplicant/src/sha2.h]: Linaro copyright to be included in the original one, without the whole license terms

File: tee-supplicant/src/sha2.c (hunk 2 above)
Comment (jforissier) [inline tee-supplicant/src/sha2.c]: Ah, sorry I forgot this one and sha2.h too. Will fix.

File: tee-supplicant/src/rpmb.c (hunk 1 above)
Comment (jforissier) [inline tee-supplicant/src/rpmb.c]: Yes, it as defined by the eMMC specification (JEDEC JESD84-B51 "6.6.22.4.4 Authenticated Data Read").

**Comment by jforissier** [comment]
Updated.

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by jforissier** [comment]
Commits squashed and review tag applied.

**Comment by jforissier** [comment]
Ping!

**Pull request merged**